### PR TITLE
Linting and formatting with Ruff

### PR DESCRIPTION
**Pull request by akx**
This PR enables linting using [Ruff](https://docs.astral.sh/ruff/) via [pre-commit](https://github.com/pre-commit/pre-commit/) (which also runs these lints in CI).

The ruff.toml configuration file is rather conservative at this point; more lint families can be added as required. The line-length of 88 characters was chosen by checking which common line-length causes the shortest `ruff format` delta.

### CODE REVIEW

**Comment by eskech** [comment]
Just a short comment. @Chroxvi  is on vacation and is back in two weeks. I really want him to do the review. Sorry for letting you wait until he is back.

**Comment by akx** [comment]
Gentle nudge (as I too am back from vacation) :)

**Comment by Chroxvi** [comment]
@akx Thanks a lot for this contribution - and sorry for not getting back to you any sooner.

A pre-commit linting setup for cotainr has been on my wishlist for quite some time, so let's get this rolling!

The [cotainr style guide](https://cotainr.readthedocs.io/en/stable/development/style_guide.html) states that we use [black](https://black.readthedocs.io/en/stable/) for code formatting, [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) for formatting docstrings, and in general adhere to [PEP8](https://peps.python.org/pep-0008/). In addition to that I have locally been using [isort](https://pycqa.github.io/isort/), [flake8](https://flake8.pycqa.org/en/latest/) and [mypy](https://mypy-lang.org/) (though mypy doesn't make much sense before we add type annotations to cotainr).

I don't have a strong opinion on any particular tool, so we can introduce ruff, if you thing that is the best suitable tool for such a pre-commit linting setup. Given the [high level of compatibility in formatting between ruff and black](https://docs.astral.sh/ruff/formatter/#black-compatibility), I think it is acceptable to switch to ruff for formatting. Would you agree that ruff is a viable solution to replace my setup with black, isort, flake8, and mypy? And do we need any additional configuration of ruff to better align with what has been the implicit standards in cotainr so far? My only concern here is to introduce too many code changes that are only related to formatting since it kind of destroys the git history.

**Comment by akx** [comment]
> I don't have a strong opinion on any particular tool, so we can introduce ruff, if you thing that is the best suitable tool for such a pre-commit linting setup. [...] Would you agree that ruff is a viable solution to replace my setup with black, isort, flake8, and mypy?

Ruff doesn't do actual type-checking, so it doesn't replace `mypy`. It can do some rudimentary annotation checks, but not actual analysis.

For the other tools, yes (disclaimer: I'm an early minor contributor to Ruff). It's a lot faster than the other tools combined (so it stays out of your way, as it were, and is more eco friendly too) :)

> And do we need any additional configuration of ruff to better align with what has been the implicit standards in cotainr so far?

More rules could be enabled in a follow-up PR to make things stricter. Ruff [can also be configured](https://docs.astral.sh/ruff/settings/#lint_pydocstyle_convention) to enforce `numpydoc` style for docstrings.

> My only concern here is to introduce too many code changes that are only related to formatting since it kind of destroys the git history.

Since ruff-format and Black are practically the same thing (aside from ruff-format doing a better job (IMO) in certain corner cases), I don't think that's an issue here.

The auto-format commit itself is quite short: https://github.com/DeiC-HPC/cotainr/pull/56/commits/bd389007398c509d06a03008d11ccdffb537f8ad

This PR could be merged with a regular merge commit, and that commit added to [`.git-blame-ignore-revs`](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/), if need be.

**Comment by akx** [comment]
I should probably split the typo fixes out of this PR, for clarity :)

**Comment by akx** [comment]
@Chroxvi Rebased post #59 merge. Anything I can do to move this (and my other PRs) forward?

**Comment by Chroxvi** [comment]
@akx Excellent, that helps a lot! At this point, I think it is entirely up to me move this forward. Sorry for not getting back to you sooner. I hope to be able to move this forward this week.

**Review by Chroxvi — CHANGES_REQUESTED** [review]
So I finally got around to digging into this PR. I still very much the idea of introducing Ruff for linting and formatting!

Having looked more into it, I am a bit more hesitant to also introduce pre-commit - see my specific review comment about this. @akx, I would like your opinion on this.

I have suggested an extended list of Ruff rules to enable based on the linting done by my editor - this may be too many rules to enable for now.

The rest is just a bunch of minor comments and suggestions for improvements.

Finally, when we get to #57, I think we should move the ruff.toml config to pyproject.toml since we are already collecting all other configuration in pyproject.toml. For now, though, lets use the ruff.toml config file.

File: .github/workflows/CI_pull_request.yml
Comment (akx) [reply .github/workflows/CI_pull_request.yml:9]: I wouldn't call it "Automatically fix", since the CI step isn't able to commit back to the branch.

Comment (Chroxvi) [reply .github/workflows/CI_pull_request.yml:9]: Is that also the case with the pre-commit/action? If so, I have misunderstood how this works. What is needed for it to automatically commit back to the branch?

File: .github/workflows/CI_pull_request.yml
Comment (akx) [reply .github/workflows/CI_pull_request.yml:16]: > However, the pre-commit/action is marked in ["maintenance-only mode"](https://github.com/pre-commit/action) which makes me hesitant to introduce it.

It's in that mode because the maintainer wants people to use pre-commit.ci, which I don't like for multiple reasons. However, the action has been proven to Just Work even if it's in "maintenance-only mode", and has done so for a long time. And as you say, if it sometime stops working, it's easy to replace. (In fact, I have a [toy project](https://github.com/akx/preco) to reimplement pre-commit in Rust, anyway...)

> We can leave out the pre-commit integration altogether

The pre-commit configuration in this PR also sets up some of pre-commit's internal hooks for maintaining more hygiene (line endings, etc), and one could set up [a whole bunch of other things there too](https://github.com/home-assistant/core/blob/dev/.pre-commit-config.yaml). (In fact, #60 was borne out of having `crate-ci/typos` in the pre-commit configuration for a while.)

Ruff's editor integration is of course fine and dandy for running Ruff. Running these lints in CI via pre-commit has proven to be a good solution for many projects I've contributed to and/or maintained.

File: cotainr/tests/tracing/test_log_dispatcher.py
Comment (akx) [reply cotainr/tests/tracing/test_log_dispatcher.py:110]: No, it was something I did, because otherwise that line would be reformatted into multiple lines in a particularly ugly way.

File: ruff.toml
Comment (akx) [reply ruff.toml:15]: Which is why I added this after rebasing :)

File: ruff.toml
Comment (akx) [reply ruff.toml:10]: With this:
```
Found 353 errors.
[*] 22 fixable with the `--fix` option (8 hidden fixes can be enabled with the `--unsafe-fixes` option).
```

File: .github/workflows/CI_pull_request.yml
Comment (Chroxvi) [reply .github/workflows/CI_pull_request.yml:9]: Alright, that's also much more in line with what I originally had in mind - no automatic fixing. But why is the `--fix` arg to ruff then included in the pre-commit config? Is that just for fixing the lint errors if run locally?

Comment (akx) [reply .github/workflows/CI_pull_request.yml:9]: > But why is the --fix arg to ruff then included in the pre-commit config? Is that just for fixing the lint errors if run locally?

Exactly. When you run `pre-commit run` (or when the Git hook runs it), you'll get the issues fixed in your work tree, and the Git hook will stop to let you stage those changes and move on.

File: cotainr/tests/tracing/test_log_dispatcher.py
Comment (akx) [reply cotainr/tests/tracing/test_log_dispatcher.py]: I added another commit before the reformatting one now that clarifies this change.

File: .github/workflows/CI_pull_request.yml
Comment (Chroxvi) [reply .github/workflows/CI_pull_request.yml:9]: That makes sense. Let's keep this approach then.

File: .github/workflows/CI_pull_request.yml
Comment (Chroxvi) [reply .github/workflows/CI_pull_request.yml:16]: Having now fully understood your setup here, I agree that this is the best approach. Let's keep the pre-commit/action.

File: ruff.toml
Comment (Chroxvi) [reply ruff.toml]: That's a lot of errors :disappointed:.  A `ruff check --statistics` gives me:
```
195	D102	[ ] undocumented-public-method
 66	D101	[ ] undocumented-public-class
 21	I001	[*] unsorted-imports
 14	D205	[ ] blank-line-after-summary
  9	D105	[ ] undocumented-magic-method
  9	D401	[ ] non-imperative-mood
  7	S603	[ ] subprocess-without-shell-equals-true
  6	D104	[ ] undocumented-public-package
  5	S607	[ ] start-process-with-partial-path
  5	D200	[*] fits-on-one-line
  3	D103	[ ] undocumented-public-function
  3	D400	[*] ends-in-period
  2	A001	[ ] builtin-variable-shadowing
  2	D404	[ ] docstring-starts-with-this
  1	S310	[ ] suspicious-url-open-usage
  1	S311	[ ] suspicious-non-cryptographic-random-usage
  1	A002	[ ] builtin-argument-shadowing
  1	D100	[ ] undocumented-public-module
  1	D106	[ ] undocumented-public-nested-class
  1	D202	[*] no-blank-line-after-function
```
So we are looking at pydocstyle, bandit, isort, builtins errors. I think we can safely include bugbear `B` and mccabe `C90`, but I suggest we leave the rest for a future PR - I think some of them may be solved by various settings, but that's probably going to take some time to sort out.

File: .github/workflows/CI_push.yml
Comment (Chroxvi) [inline .github/workflows/CI_push.yml:15]:
```suggestion
          python-version: "3.x"
```

File: cotainr/cli.py
Comment (Chroxvi) [reply cotainr/cli.py]: I guess I was a little too fast here. It seems that this suggested change has made quite a lot of tests fail since we actually don't expect `CotainrSubcommand` subclasses to implement the `add_arguments()` method if they take no arguments. I think we should stick with your proposal of just adding a `# noqa: B027`.

File: .github/workflows/CI_pull_request.yml
Comment (Chroxvi) [inline .github/workflows/CI_pull_request.yml:9]:
```suggestion
  lint-and-format:
    name: Check for linting and formatting errors
```

**Comment by Chroxvi** [comment]
@akx I think the main missing part here is handling my comments related to adding more details about this linting setup to the documentation. Would you be willing to add these? If not, I am also very happy to merge this PR once the details in the other few remaining comments have been sorted out.

**Comment by akx** [comment]
> I think the main missing part here is handling my comments related to adding more details about this linting setup to the documentation. Would you be willing to add these?

Done deal. To keep this more reviewable, I can follow this up after merging with another PR that adds more `select`s as discussed in https://github.com/DeiC-HPC/cotainr/pull/56#discussion_r1750032345

File: doc/development/style_guide.rst
Comment (akx) [reply doc/development/style_guide.rst:8]: I think the configuration file's location is an implementation detail. `pytest.ini` being the pytest configuration file is not documented either, etc. :)

**Comment by Chroxvi** [comment]
Excellet! I'll merge this now.

@akx Thanks for a lot for hanging in here and for all the valuable discussions around best practices for using Ruff, pre-commit, and GitHub actions.

**Review by Chroxvi — APPROVED** [review]
Approved!

**Pull request merged**
